File: frontend/lib/models/permission.dart
/// Permissions in the CareConnect system
/// Must match backend Permission enum exactly (26 total)
enum Permission {
  // User Management (3)
  viewAllUsers,
  manageUsers,
  assignRoles,

  // Patient Management (4)
  viewAssignedPatients,
  createPatients,
  updatePatients,
  deletePatients,

  // Health Data (3)
  viewHealthData,
  recordHealthData,
  exportHealthData,

  // Task Management (5)
  viewTasks,
  createTasks,
  updateTasks,
  deleteTasks,
  completeTasks,

  // Medication Management (2)
  viewMedications,
  manageMedications,

  // Analytics & Reports (2)
  viewAnalytics,
  exportReports,

  // Messaging (2)
  viewMessages,
  sendMessages,

  // Billing (2)
  viewBilling,
  manageSubscriptions,

  // System (2)
  manageSystemSettings,
  viewAuditLogs;

  /// Convert from backend string (SCREAMING_SNAKE_CASE) to Dart enum
  static Permission fromString(String permissionString) {
    switch (permissionString.toUpperCase()) {
      // User Management
      case 'VIEW_ALL_USERS':
        return Permission.viewAllUsers;
      case 'MANAGE_USERS':
        return Permission.manageUsers;
      case 'ASSIGN_ROLES':
        return Permission.assignRoles;

      // Patient Management
      case 'VIEW_ASSIGNED_PATIENTS':
        return Permission.viewAssignedPatients;
      case 'CREATE_PATIENTS':
        return Permission.createPatients;
      case 'UPDATE_PATIENTS':
        return Permission.updatePatients;
      case 'DELETE_PATIENTS':
        return Permission.deletePatients;

      // Health Data
      case 'VIEW_HEALTH_DATA':
        return Permission.viewHealthData;
      case 'RECORD_HEALTH_DATA':
        return Permission.recordHealthData;
      case 'EXPORT_HEALTH_DATA':
        return Permission.exportHealthData;

      // Task Management
      case 'VIEW_TASKS':
        return Permission.viewTasks;
      case 'CREATE_TASKS':
        return Permission.createTasks;
      case 'UPDATE_TASKS':
        return Permission.updateTasks;
      case 'DELETE_TASKS':
        return Permission.deleteTasks;
      case 'COMPLETE_TASKS':
        return Permission.completeTasks;

      // Medication Management
      case 'VIEW_MEDICATIONS':
        return Permission.viewMedications;
      case 'MANAGE_MEDICATIONS':
        return Permission.manageMedications;

      // Analytics & Reports
      case 'VIEW_ANALYTICS':
        return Permission.viewAnalytics;
      case 'EXPORT_REPORTS':
        return Permission.exportReports;

      // Messaging
      case 'VIEW_MESSAGES':
        return Permission.viewMessages;
      case 'SEND_MESSAGES':
        return Permission.sendMessages;

      // Billing
      case 'VIEW_BILLING':
        return Permission.viewBilling;
      case 'MANAGE_SUBSCRIPTIONS':
        return Permission.manageSubscriptions;

      // System
      case 'MANAGE_SYSTEM_SETTINGS':
        return Permission.manageSystemSettings;
      case 'VIEW_AUDIT_LOGS':
        return Permission.viewAuditLogs;

      default:
        throw ArgumentError('Unknown permission: $permissionString');
    }
  }

  /// Convert to backend format (SCREAMING_SNAKE_CASE)
  String toBackendString() {
    switch (this) {
      // User Management
      case Permission.viewAllUsers:
        return 'VIEW_ALL_USERS';
      case Permission.manageUsers:
        return 'MANAGE_USERS';
      case Permission.assignRoles:
        return 'ASSIGN_ROLES';

      // Patient Management
      case Permission.viewAssignedPatients:
        return 'VIEW_ASSIGNED_PATIENTS';
      case Permission.createPatients:
        return 'CREATE_PATIENTS';
      case Permission.updatePatients:
        return 'UPDATE_PATIENTS';
      case Permission.deletePatients:
        return 'DELETE_PATIENTS';

      // Health Data
      case Permission.viewHealthData:
        return 'VIEW_HEALTH_DATA';
      case Permission.recordHealthData:
        return 'RECORD_HEALTH_DATA';
      case Permission.exportHealthData:
        return 'EXPORT_HEALTH_DATA';

      // Task Management
      case Permission.viewTasks:
        return 'VIEW_TASKS';
      case Permission.createTasks:
        return 'CREATE_TASKS';
      case Permission.updateTasks:
        return 'UPDATE_TASKS';
      case Permission.deleteTasks:
        return 'DELETE_TASKS';
      case Permission.completeTasks:
        return 'COMPLETE_TASKS';

      // Medication Management
      case Permission.viewMedications:
        return 'VIEW_MEDICATIONS';
      case Permission.manageMedications:
        return 'MANAGE_MEDICATIONS';

      // Analytics & Reports
      case Permission.viewAnalytics:
        return 'VIEW_ANALYTICS';
      case Permission.exportReports:
        return 'EXPORT_REPORTS';

      // Messaging
      case Permission.viewMessages:
        return 'VIEW_MESSAGES';
      case Permission.sendMessages:
        return 'SEND_MESSAGES';

      // Billing
      case Permission.viewBilling:
        return 'VIEW_BILLING';
      case Permission.manageSubscriptions:
        return 'MANAGE_SUBSCRIPTIONS';

      // System
      case Permission.manageSystemSettings:
        return 'MANAGE_SYSTEM_SETTINGS';
      case Permission.viewAuditLogs:
        return 'VIEW_AUDIT_LOGS';
    }
  }

  /// Get display name for UI (converts camelCase to Title Case)
  String get displayName {
    final name = toString().split('.').last;
    final result = name.replaceAllMapped(
      RegExp(r'([A-Z])'),
      (match) => ' ${match.group(0)}',
    );
    return result[0].toUpperCase() + result.substring(1);
  }

  /// Get permission description
  String get description {
    switch (this) {
      // User Management
      case Permission.viewAllUsers:
        return 'View all users in the system';
      case Permission.manageUsers:
        return 'Create, update, and delete users';
      case Permission.assignRoles:
        return 'Assign and modify user roles';

      // Patient Management
      case Permission.viewAssignedPatients:
        return 'View assigned patient information';
      case Permission.createPatients:
        return 'Create new patient records';
      case Permission.updatePatients:
        return 'Update existing patient information';
      case Permission.deletePatients:
        return 'Delete patient records';

      // Health Data
      case Permission.viewHealthData:
        return 'View patient health data';
      case Permission.recordHealthData:
        return 'Record and update health metrics';
      case Permission.exportHealthData:
        return 'Export health data reports';

      // Task Management
      case Permission.viewTasks:
        return 'View tasks and care plans';
      case Permission.createTasks:
        return 'Create new tasks';
      case Permission.updateTasks:
        return 'Update existing tasks';
      case Permission.deleteTasks:
        return 'Delete tasks';
      case Permission.completeTasks:
        return 'Mark tasks as complete';

      // Medication Management
      case Permission.viewMedications:
        return 'View medication schedules';
      case Permission.manageMedications:
        return 'Add, update, and remove medications';

      // Analytics & Reports
      case Permission.viewAnalytics:
        return 'View analytics and insights';
      case Permission.exportReports:
        return 'Generate and export reports';

      // Messaging
      case Permission.viewMessages:
        return 'View messages and communications';
      case Permission.sendMessages:
        return 'Send messages to other users';

      // Billing
      case Permission.viewBilling:
        return 'View billing and payment information';
      case Permission.manageSubscriptions:
        return 'Manage subscriptions and plans';

      // System
      case Permission.manageSystemSettings:
        return 'Configure system settings';
      case Permission.viewAuditLogs:
        return 'View system audit logs';
    }
  }

  /// Check if this permission is admin-only
  bool get isAdminOnly {
    return this == Permission.viewAllUsers ||
        this == Permission.manageUsers ||
        this == Permission.assignRoles ||
        this == Permission.deletePatients ||
        this == Permission.manageSystemSettings ||
        this == Permission.viewAuditLogs;
  }

  /// Get permission category
  PermissionCategory get category {
    if ([viewAllUsers, manageUsers, assignRoles].contains(this)) {
      return PermissionCategory.userManagement;
    }
    if ([viewAssignedPatients, createPatients, updatePatients, deletePatients]
        .contains(this)) {
      return PermissionCategory.patientManagement;
    }
    if ([viewHealthData, recordHealthData, exportHealthData].contains(this)) {
      return PermissionCategory.healthData;
    }
    if ([viewTasks, createTasks, updateTasks, deleteTasks, completeTasks]
        .contains(this)) {
      return PermissionCategory.taskManagement;
    }
    if ([viewMedications, manageMedications].contains(this)) {
      return PermissionCategory.medicationManagement;
    }
    if ([viewAnalytics, exportReports].contains(this)) {
      return PermissionCategory.analytics;
    }
    if ([viewMessages, sendMessages].contains(this)) {
      return PermissionCategory.messaging;
    }
    if ([viewBilling, manageSubscriptions].contains(this)) {
      return PermissionCategory.billing;
    }
    return PermissionCategory.system;
  }
}

/// Permission categories for grouping and display
enum PermissionCategory {
  userManagement,
  patientManagement,
  healthData,
  taskManagement,
  medicationManagement,
  analytics,
  messaging,
  billing,
  system;

  String get displayName {
    switch (this) {
      case PermissionCategory.userManagement:
        return 'User Management';
      case PermissionCategory.patientManagement:
        return 'Patient Management';
      case PermissionCategory.healthData:
        return 'Health Data';
      case PermissionCategory.taskManagement:
        return 'Task Management';
      case PermissionCategory.medicationManagement:
        return 'Medication Management';
      case PermissionCategory.analytics:
        return 'Analytics & Reports';
      case PermissionCategory.messaging:
        return 'Messaging';
      case PermissionCategory.billing:
        return 'Billing';
      case PermissionCategory.system:
        return 'System';
    }
  }
}

File: backend/core/src/main/java/com/careconnect/security/Permission.java
package com.careconnect.security;

/**
 * Enumeration of all permissions in the CareConnect system.
 *
 * Permissions represent fine-grained actions that users can perform.
 * Each permission is assigned to one or more roles via the RolePermissionService.
 * 
 * Usage:
 *   - Check if user has permission: user.hasPermission(Permission.CREATE_TASKS)
 *   - In controllers: authorizationService.requirePermission(user, Permission.VIEW_HEALTH_DATA)
 * 
 * @author CareConnect Team
 * @version 1.0
 */
public enum Permission {
    
    // ========== User Management Permissions ==========
    
    /**
     * View all users in the system (admin only).
     * Allows access to complete user directory.
     */
    VIEW_ALL_USERS("View all users in the system"),
    
    /**
     * Create, update, and delete user accounts.
     * Includes modifying user profile information.
     */
    MANAGE_USERS("Create, update, delete users"),
    
    /**
     * Change user roles and permissions.
     * Critical permission - admin only.
     */
    ASSIGN_ROLES("Change user roles"),
    
    
    // ========== Patient Management Permissions ==========
    
    /**
     * View all patients in the system (admin only).
     * Unrestricted access to all patient records.
     */
    VIEW_ALL_PATIENTS("View all patients (admin only)"),
    
    /**
     * View patients assigned to the current user.
     * Caregivers see only their assigned patients.
     */
    VIEW_ASSIGNED_PATIENTS("View patients assigned to you"),
    
    /**
     * Create new patient profiles.
     * Includes entering demographic and medical information.
     */
    CREATE_PATIENTS("Create new patient profiles"),
    
    /**
     * Update existing patient information.
     * Modify demographics, medical history, contact info.
     */
    UPDATE_PATIENTS("Update patient information"),
    
    /**
     * Delete or archive patient profiles.
     * Permanently removes or soft-deletes patient records.
     */
    DELETE_PATIENTS("Delete patient profiles"),
    
    
    // ========== Task Management Permissions ==========
    
    /**
     * Create tasks and care schedules for patients.
     * Includes one-time and recurring tasks.
     */
    CREATE_TASKS("Create tasks for patients"),
    
    /**
     * View tasks and schedules.
     * Filtered by patient access permissions.
     */
    VIEW_TASKS("View tasks"),
    
    /**
     * Update task details, schedules, and priorities.
     * Modify existing task information.
     */
    UPDATE_TASKS("Update task details"),
    
    /**
     * Delete tasks and schedules.
     * Remove tasks from the system.
     */
    DELETE_TASKS("Delete tasks"),
    
    /**
     * Mark tasks as complete.
     * Patients can complete their own tasks, caregivers can complete any assigned task.
     */
    COMPLETE_TASKS("Mark tasks as complete"),
    
    
    // ========== Health Data Permissions ==========
    
    /**
     * View patient health metrics and vital signs.
     * Includes blood pressure, heart rate, glucose, weight, etc.
     */
    VIEW_HEALTH_DATA("View patient health metrics"),
    
    /**
     * Record new health measurements.
     * Add new vital signs and health data points.
     */
    RECORD_HEALTH_DATA("Record health measurements"),
    
    /**
     * Export health data reports.
     * Generate CSV, PDF, or other formats for download.
     */
    EXPORT_HEALTH_DATA("Export health data reports"),


    // ========== Medication Permissions ==========

    /**
     * View medication information for patients.
     * Access to medication lists and prescription details.
     */
    VIEW_MEDICATIONS("View medication information"),

    /**
     * Manage medications for patients.
     * Add, update, and remove medication records.
     */
    MANAGE_MEDICATIONS("Manage patient medications"),


    // ========== Billing & Subscription Permissions ==========
    
    /**
     * View billing information and invoice history.
     * Access to payment records and subscription status.
     */
    VIEW_BILLING("View billing information"),
    
    /**
     * Manage subscription plans and payments.
     * Subscribe, upgrade, downgrade, cancel subscriptions.
     */
    MANAGE_SUBSCRIPTIONS("Manage subscription plans"),
    
    
    // ========== Communication Permissions ==========
    
    /**
     * Send messages to care circle members.
     * Includes text messages, images, and file attachments.
     */
    SEND_MESSAGES("Send messages to care circle"),
    
    /**
     * View messages and conversation history.
     * Access to message threads and communications.
     */
    VIEW_MESSAGES("View messages"),
    
    
    // ========== Analytics & Reporting Permissions ==========
    
    /**
     * View analytics dashboards and metrics.
     * Access to QuickSight dashboards and data visualizations.
     */
    VIEW_ANALYTICS("View analytics dashboards"),
    
    /**
     * Export reports and analytics data.
     * Generate and download analytics reports.
     */
    EXPORT_REPORTS("Export reports and data"),
    
    
    // ========== AI & Assistant Permissions ==========
    
    /**
     * Use AI assistant features.
     * Access to AI-powered health insights and recommendations.
     */
    USE_AI_FEATURES("Use AI assistant features"),
    
    
    // ========== Device Integration Permissions ==========
    
    /**
     * Connect and manage wearable devices.
     * Link Fitbit, Apple Health, Google Fit devices.
     */
    MANAGE_DEVICES("Connect and manage wearable devices"),
    
    
    // ========== Notification & Audit Permissions ==========
    
    /**
     * Manage notification preferences and settings.
     * Configure alert types, delivery methods, and schedules.
     */
    MANAGE_NOTIFICATIONS("Manage notification preferences"),
    
    /**
     * View system audit logs and access history.
     * Access to security logs and compliance reports (admin only).
     */
    VIEW_AUDIT_LOGS("View system audit logs");
    
    
    // ========== Instance Variables ==========
    
    /**
     * Human-readable description of what this permission allows.
     */
    private final String description;
    
    
    // ========== Constructor ==========
    
    /**
     * Private constructor called automatically for each enum constant.
     * 
     * @param description Human-readable description of the permission
     */
    Permission(String description) {
        this.description = description;
    }

    // ========== Public Methods ==========

    /**
     * Gets the human-readable description of this permission.
     *
     * @return Description explaining what this permission allows
     */
    public String getDescription() {
        return description;
    }
    
    /**
     * Gets the permission name formatted for display.
     * Converts CREATE_TASKS to "Create Tasks" with each word capitalized.
     * 
     * @return Formatted permission name with proper capitalization
     * 
     * @example
     * Permission.CREATE_TASKS.getDisplayName() returns "Create Tasks"
     * Permission.VIEW_HEALTH_DATA.getDisplayName() returns "View Health Data"
     */
    public String getDisplayName() {
        // Split on underscores: CREATE_TASKS -> ["CREATE", "TASKS"]
        String[] words = this.name().split("_");
        StringBuilder result = new StringBuilder();
        
        for (String word : words) {
            if (result.length() > 0) {
                result.append(" ");
            }
            // Capitalize first letter, lowercase the rest
            // "CREATE" -> "Create"
            result.append(Character.toUpperCase(word.charAt(0)))
                  .append(word.substring(1).toLowerCase());
        }
        
        return result.toString();
    }
    
    /**
     * Checks if this is an admin-only permission.
     * 
     * @return true if permission is restricted to admins
     */
    public boolean isAdminOnly() {
        return this == VIEW_ALL_USERS || 
               this == MANAGE_USERS || 
               this == ASSIGN_ROLES || 
               this == VIEW_ALL_PATIENTS || 
               this == DELETE_PATIENTS ||
               this == VIEW_AUDIT_LOGS;
    }
    
    /**
     * Returns a string representation of this permission.
     * 
     * @return Permission name and description
     */
    @Override
    public String toString() {
        return this.name() + ": " + description;
    }
}